MapboxNavigation/MapboxSpeechSynthesizer.swift

```diff
@@ -0,0 +1,242 @@
+
+import AVFoundation
+import MapboxCoreNavigation
+import MapboxDirections
+import MapboxSpeech
+
+/**
+ `SpeechSynthesizing` implementation, using `MapboxSpeech` framework. Uses pre-caching mechanism for upcoming instructions.
+ */
+open class MapboxSpeechSynthesizer: NSObject, SpeechSynthesizing {
+    
+    // MARK: - Properties
+    
+    public weak var delegate: SpeechSynthesizingDelegate?
+    public var muted: Bool = false {
+        didSet {
+            updatePlayerVolume(audioPlayer)
+        }
+    }
+    public var volume: Float = 1.0 {
+        didSet {
+            audioPlayer?.volume = volume
+        }
+    }
+    public var isSpeaking: Bool {
+        return audioPlayer?.isPlaying ?? false
+    }
+    public var locale: Locale = Locale.autoupdatingCurrent
+    
+    public let stepsAheadToCache: Int = 3
+    
+    /**
+     An `AVAudioPlayer` through which spoken instructions are played.
+     */
+    public var audioPlayer: AVAudioPlayer?
+    
+    private var cache: BimodalDataCache
+    private var speech: SpeechSynthesizer
+    private var audioTask: URLSessionDataTask?
+    
+    private var previousInstruction: SpokenInstruction?
+    
+    // MARK: - Lifecycle
+    
+    public init(_ accessToken: String? = nil, host: String? = nil) {
+        self.cache = DataCache()
+        self.speech = SpeechSynthesizer(accessToken: accessToken, host: host)
+    }
+    
+    deinit {        
+        deinitAudioPlayer()
+    }
+    
+    // MARK: - Methods
+    
+    open func prepareIncomingSpokenInstructions(_ instructions: [SpokenInstruction]) {
+        instructions
+            .prefix(stepsAheadToCache)
+            .forEach {
+                if !hasCachedSpokenInstructionForKey($0.ssmlText, with: locale) {
+                    downloadAndCacheSpokenInstruction(instruction: $0)
+                }
+        }
+    }
+    
+    open func speak(_ instruction: SpokenInstruction, during legProgress: RouteLegProgress) {
+        if let data = cachedDataForKey(instruction.ssmlText, with: locale) {
+            safeDuckAudio(instruction: instruction)
+            speak(instruction: instruction,
+                  instructionData: data)
+        }
+        else {
+            fetchAndSpeak(instruction: instruction)
+        }
+    }
+    
+    open func stopSpeaking() {
+        audioPlayer?.stop()
+    }
+    
+    open func interruptSpeaking() {
+        audioPlayer?.stop()
+    }
+    
+    // MARK: - Private Methods
+    
+    /**
+     Fetches and plays an instruction.
+     */
+    private func fetchAndSpeak(instruction: SpokenInstruction){
+        audioTask?.cancel()
+        
+        let modifiedInstruction = delegate?.speechSynthesizer(self, willSpeak: instruction) ?? instruction
+        let ssmlText = modifiedInstruction.ssmlText
+        let options = SpeechOptions(ssml: ssmlText)
+        options.locale = locale
```

Comment:
By default, this locale should be the `speechLocale` from the route response, if present, to ensure that we use the voice that the text was designed for. Otherwise, the fallback to `autoupdatingCurrent` is fine, but that shouldn’t be the default. I think this context can be plumbed through the `speak(_:during:)` call.